oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,271 @@
+                val message = OASISQueryDTO.newBuilder()
+                        .setEvaluationRequest(OASISQueryDTO.SimulationEvaluationRequest.newBuilder()
+                                .setName(simName)
+                                .putAllInputVector(inputVector)
+                        )
+                        .build()
+
+                sim.input.onNext(message)
+                val result = select<SimResult> {
+                    sim.output.onReceive { SimResult.Success(it.name, it.outputVectorMap) }
+                    sim.error.onReceive { SimResult.Failure(it.name, it.exception) }
+                    if (sim.timeOut != null) {
+                        onTimeout(sim.timeOut.toMillis()) {
+                            //Timed out
+                            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(simName)
+                            ).build()
+                            sim.input.onNext(message)
+                            select {
+                                sim.output.onReceive { SimResult.TimeOut(it.name) }
+                                sim.error.onReceive { SimResult.TimeOutFailure(it.name, it.exception) }
+                            }
+                        }
+                    }
+                }
+
+                val newResultMessage: Message
+                val resultMessage: Result
+                when (result) {
+                    is SimResult.Success -> {
+                        newResultMessage = Message(result.name, "Evaluation Succeed: Result [${result.result}]")
+                        resultMessage = Result(result.name, "Success", inputVector.toString(), result.result.toString())
+                    }
+                    is SimResult.Failure -> {
+                        newResultMessage = Message(result.name, "Evaluation Failed: Due to\n${result.exception}")
+                        resultMessage = Result(result.name, "Error", inputVector.toString(), "Failure: \n${result.exception}")
+                    }
+                    is SimResult.TimeOut -> {
+                        newResultMessage = Message(result.name, "Evaluation Canceled to timed out after ${sim.timeOut!!.toMillis()} millisecond")
+                        resultMessage = Result(result.name, "Canceled", inputVector.toString(), "Timeout")
+                    }
+                    is SimResult.TimeOutFailure -> {
+                        newResultMessage = Message(result.name, "Evaluation try to cancel Timed out after ${sim.timeOut!!.toMillis()} millisecond but Failed: Due to\n${result.exception}")
+                        resultMessage = Result(result.name, "Cancel Failed", inputVector.toString(), "Cancellation failed:\n${result.exception}")
+                    }
+                }
+                modelService.addMessage(newResultMessage)
+                modelService.addResult(resultMessage)
```

Comment:
I really want to decouple you from manual logging systems like this. They are increadibly verbose. 

Rather than have to manually call `addMessage` every time you make a result, what about putting an interceptor that intercepts outbound messages and calls `addMessage()` with them?

Reply:
No, you are right, I don't really need to "do my own logging" especially using addMessage() add result is to replace this and show result in a better place.